src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,1089 @@
+{
+    // Count the number of matching "DNS/SRP Anycast" service entries
+    // in the Network Data (the match requires the entry to use same
+    // "sequence number" value). We prefer the entries associated with
+    // smaller RLCO16.
+
+    Service::DnsSrpAnycast::ServiceData serviceData(mInfo.GetSequenceNumber());
+    const ServiceTlv *                  serviceTlv = nullptr;
+
+    while ((serviceTlv = Get<Leader>().FindNextMatchingService(serviceTlv, Service::kThreadEnterpriseNumber,
+                                                               reinterpret_cast<const uint8_t *>(&serviceData),
+                                                               serviceData.GetLength())) != nullptr)
+    {
+        TlvIterator      subTlvIterator(*serviceTlv);
+        const ServerTlv *serverSubTlv;
+
+        while ((serverSubTlv = subTlvIterator.Iterate<ServerTlv>()) != nullptr)
+        {
+            aNumEntries++;
+
+            if (IsPreferred(serverSubTlv->GetServer16()))
+            {
+                aNumPreferredEntries++;
+            }
+        }
+    }
+}
+
+void Publisher::DnsSrpServiceEntry::CountUnicastEntries(uint8_t &aNumEntries, uint8_t &aNumPreferredEntries) const
+{
+    // Count the number of "DNS/SRP Unicast" service entries in
+    // the Network Data.
+
+    const ServiceTlv *serviceTlv = nullptr;
+
+    while ((serviceTlv = Get<Leader>().FindNextMatchingService(
+                serviceTlv, Service::kThreadEnterpriseNumber, &Service::DnsSrpUnicast::kServiceData,
+                sizeof(Service::DnsSrpUnicast::kServiceData))) != nullptr)
+    {
+        TlvIterator      subTlvIterator(*serviceTlv);
+        const ServerTlv *serverSubTlv;
+
+        while (((serverSubTlv = subTlvIterator.Iterate<ServerTlv>())) != nullptr)
+        {
+            if (serviceTlv->GetServiceDataLength() >= sizeof(Service::DnsSrpUnicast::ServiceData))
+            {
+                aNumEntries++;
+
+                // Generally, we prefer entries where the SRP/DNS server
+                // address/port info is included in the service TLV data
+                // over the ones where the info is included in the
+                // server TLV data (i.e., we prefer infra-provided
+                // SRP/DNS entry over a BR local one using ML-EID). If
+                // our entry itself uses the service TLV data, then we
+                // prefer based on the associated RLOC16.
+
+                if (GetType() == kTypeUnicast)
+                {
+                    if (IsPreferred(serverSubTlv->GetServer16()))
+                    {
+                        aNumPreferredEntries++;
+                    }
+                }
+                else
+                {
+                    aNumPreferredEntries++;
+                }
+            }
+
+            if (serverSubTlv->GetServerDataLength() >= sizeof(Service::DnsSrpUnicast::ServerData))
```

Comment:
What about Server TLVs that don't include the IP address? Do we support such format?

Reply:
Yes. Definitely (we use this when we publish using "ML-EID". 

In such a case, we only count the entry if the enclosing service TLV data contains the IP address. That's the first `if` here in this loop.

Basically here in the code we are iterating over all Server sub-TLVs:
- We first check if the enclosing Service TLV does contain IPv6 address, if so, we count it as one entry (each server TLV represents one RLOC (one BR) supporting this address). 
- Then we also check if the "server sub-TLV" has an address embedded in its data and if it does, we count it as an entry as well.

Reply:
So we are not counting the SRP servers that include an IP address in neither `Service TLV` nor `Server TLV`? I think such format is okay and am wondering why we are not counting it as an unicast entry.

Reply:
That's correct. The "Unicast Service Entry" definition allows the address info (address & port) to be either included as part of "service data" or "server data". If both are empty, then it is not a valid entry (other nodes cannot use such an entry to determine the address and port info of the server). And we don't count them.

Reply:
I thought we can construct the server address with RLOC16 in the Server TLV.Is it because the RLOC address is not stable that we don't support such format?

Reply:
Using RLOC16 from Server sub-TLV to determine the server address is not supported since it won't work in all cases:
- MTD devices that don't get the full network data won't see or know the RLOC16 of entries:
   -  Parent will replace  the RLOC16 in all TLV entries with `0xfffe` (or ALOCs) before sending it to such a child
   - This is done from `NetworkData::RemoveTemporaryData()` methods (and is what Thread spec requires).
- We also need the port number (we could consider assuming a fixed/known port number but the draft spec requires port number to be included for unicast address model).

Reply:
Thanks! that's clear.